File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;

        int[] rightArr = new int[n];

        int right = 1;

        for (int i = n - 1; i >= 0; i--) {
            right *= nums[i];
            rightArr[i] = right;
        }

        int[] ansArr = new int[n];

        int left = 1;

        for (int i = 0; i < n - 1; i++) {
            int pro = left * rightArr[i + 1];

            ansArr[i] = pro;

            left *= nums[i];
        }

        ansArr[n - 1] = left;

        return ansArr;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
    let len = nums.length;

    let right = [];
    right.length = len;

    let pro = 1;

    for (let i = len - 1; i >= 0; i--) {
        pro = pro * nums[i];
        right[i] = pro;
    }

    let ans = [];
    ans.length = len;

    let left = 1;

    let i = 0;

    while (i < len - 1) {
        let pro = left * right[i + 1];
        ans[i] = pro;
        left = left * nums[i];
        i++;
    }

    ans[len -1] = left;

    return ans;
};